src/front/wgsl.rs

```diff
@@ -456,6 +458,27 @@ impl Parser {
         }
     }
 
+    fn deconstruct_composite_type(
+        type_arena: &mut Arena<crate::Type>,
+        ty: Handle<crate::Type>,
+        index: usize,
+    ) -> Result<Handle<crate::Type>, Error<'static>> {
+        let ty = match &type_arena[ty].inner {
```

Comment:
let's not match on a reference, please

Reply:
How should I do this then? If I match on the value, it'll try to move it and it can't.

Reply:
Use `ref` in pattern matching instead, where the data is not `Copy`:
```rust
let ty = match type_arena[ty].inner {
 ...
  crate::TypeInner::Struct { ref members } => {..}
};
```

Reply:
There's some ownership issues with that, but I'll work around them. For future reference, why is matching on a reference discouraged?

Reply:
`ref` is precisely what you get behind the sugar of matching on a reference, and that's why I don't like it.
Sorry, it could be one of my pet peeves. I'm used to Rust that doesn't even allow one to match on a reference, and you had to write out `ref` where applicable. That has the benefit of you knowing precisely what's going on there upon reading some code, and it allows you to use `ref` even on `Copy` types in case they are big, and you don't really want them to be copied anywhere.

TL;DR: there shouldn't be any more ownership issues with `ref` than you'd get with matching on a reference

Reply:
Weird, maybe I was doing it wrong, but it was borrowing the `type_arena` through the whole match expression when I need to use it mutably. Whereas, matching on references didn't do that.

Reply:
if you need mutability, you do `ref mut` instead of `ref`